### ISSUE TITLE
Removing NodeJS Console Window

### ISSUE THREAD

**Issue by hansooms999** [JeringTech/Javascript.NodeJS#92]
Hello there,

I'm unsure if this is an issue or a lack of understanding on my own part but I can't find anything mentioning it in the documentation. 

I use InvokeFromFileAsync to execute a NodeJS script in a C# Windows Forms application. It works great, however there is a Node console window popup every time the script is executed. It doesn't really add anything to my UI and I'm unsure how to execute the process completely in the background so this popup doesn't occur. 

<img width="997" alt="Screen Shot 2021-01-12 at 11 37 43 AM" src="https://user-images.githubusercontent.com/43038327/104344195-d6df0880-54ca-11eb-8f96-6c8be5f7e925.png">

Any help is appreciated,

Thank you!

**Comment by JeremyTCD** [comment]
Hey! We got a direct email on this issue 3 weeks back. Out of curiosity, was that you? Asking to ascertain if this is the second occurrence of this issue. That would strongly suggest the problem is with our library. The console window should not open.

Anyway I was unable to reproduce the problem that time. Is it possible for you to create a minimal reproduction? 
So create a new project with the same configuration as yours and call `InvokeFromFileAsync` on a dummy js file. Then push the repro to Github.  

Would be more than happy to look at it and figure this out.

**Comment by hansooms999** [comment]
No the email wasn't me, I'll let you know when I get a sample repo to showcase the problem.

Thank you for the quick response!

**Comment by JeremyTCD** [comment]
After some digging, I think the following change to this library may fix the issue:

https://github.com/JeringTech/Javascript.NodeJS/blob/b25a9edf67be48b4fb74f8498117bd61610e8b54/src/NodeJS/NodeJSServiceImplementations/OutOfProcess/NodeJSProcessFactory.cs#L37-L45

Changed to:

```cs
var startInfo = new ProcessStartInfo("node")
{
    Arguments = $"{_nodeJSProcessOptions.NodeAndV8Options} -e \"{nodeServerScript}\" -- --parentPid {currentProcessPid} --port {_nodeJSProcessOptions.Port}",
    UseShellExecute = false,
    RedirectStandardInput = true,
    RedirectStandardOutput = true,
    RedirectStandardError = true,
    WorkingDirectory = _nodeJSProcessOptions.ProjectPath,
    /*** Possible solution for NodeJS console window issue ***/
    CreateNoWindow = true
};
```

Note 100% sure if this fixes the issue, but worth trying. If you could test this change out I'd appreciate it. Steps:

1. Add the following `CustomNodeJSProcessFactory .cs` file to your project. It's identical to the [original class that creates NodeJS processes](https://github.com/JeringTech/Javascript.NodeJS/blob/master/src/NodeJS/NodeJSServiceImplementations/OutOfProcess/NodeJSProcessFactory.cs) except it sets the property mentioned above. You might have to add several packages to your project temporarily and you'll have to enter your namespace:
    ```cs
    using Jering.Javascript.NodeJS;
    using Microsoft.Extensions.Options;
    using System;
    using System.Diagnostics;
    using System.Text;
    
    namespace <your namespace>
    {
        /// <summary>
        /// Custom <see cref="INodeJSProcessFactory"/> implementation.
        /// </summary>
        public class CustomNodeJSProcessFactory : INodeJSProcessFactory
        {
            private readonly NodeJSProcessOptions _nodeJSProcessOptions;
    
            /// <summary>
            /// Creates a <see cref="NodeJSProcessFactory"/>.
            /// </summary>
            /// <param name="optionsAccessor"></param>
            public CustomNodeJSProcessFactory(IOptions<NodeJSProcessOptions> optionsAccessor)
            {
                _nodeJSProcessOptions = optionsAccessor?.Value ?? new NodeJSProcessOptions();
            }
    
            /// <inheritdoc />
            public INodeJSProcess Create(string serverScript)
            {
                Console.WriteLine("Creating process using custom NodeJSProcessFactory");
    
                ProcessStartInfo startInfo = CreateStartInfo(serverScript);
    
                return new NodeJSProcess(CreateProcess(startInfo));
            }
    
            internal ProcessStartInfo CreateStartInfo(string nodeServerScript)
            {
                nodeServerScript = EscapeCommandLineArg(nodeServerScript); // TODO can we escape before embedding? Would avoid an allocation every time we start a NodeJS process.
    
                int currentProcessPid = Process.GetCurrentProcess().Id;
                var startInfo = new ProcessStartInfo("node")
                {
                    Arguments = $"{_nodeJSProcessOptions.NodeAndV8Options} -e \"{nodeServerScript}\" -- --parentPid {currentProcessPid} --port {_nodeJSProcessOptions.Port}",
                    UseShellExecute = false,
                    RedirectStandardInput = true,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    WorkingDirectory = _nodeJSProcessOptions.ProjectPath,
                    /*** Possible solution for NodeJS console window issue ***/
                    CreateNoWindow = true
                };
    
                // Append environment Variables
                if (_nodeJSProcessOptions.EnvironmentVariables != null)
                {
                    foreach (var envVarKey in _nodeJSProcessOptions.EnvironmentVariables.Keys)
                    {
                        string envVarValue = _nodeJSProcessOptions.EnvironmentVariables[envVarKey];
                        if (envVarValue != null)
                        {
                            startInfo.Environment[envVarKey] = envVarValue;
                        }
                    }
                }
    
                return startInfo;
            }
    
            internal Process CreateProcess(ProcessStartInfo startInfo)
            {
                try
                {
                    Process process = Process.Start(startInfo);
    
                    // On Mac at least, a killed child process is left open as a zombie until the parent
                    // captures its exit code. We don't need the exit code for this process, and don't want
                    // to use process.WaitForExit() explicitly (we'd have to block the thread until it really
                    // has exited), but we don't want to leave zombies lying around either. It's sufficient
                    // to use process.EnableRaisingEvents so that .NET will grab the exit code and let the
                    // zombie be cleaned away without having to block our thread.
                    process.EnableRaisingEvents = true;
    
                    return process;
                }
                catch (Exception exception)
                {
                    throw new InvalidOperationException("Failed to start NodeJS process", exception);
                }
            }
    
            internal string EscapeCommandLineArg(string arg)
            {
                var stringBuilder = new StringBuilder();
                int slashSequenceLength = 0;
                for (int i = 0; i < arg.Length; i++)
                {
                    char currentChar = arg[i];
    
                    if (currentChar == '\\')
                    {
                        slashSequenceLength++;
    
                        // If the last character in the argument is \, it must be escaped, together with any \ that immediately preceed it.
                        // This prevents situations like: SomeExecutable.exe "SomeArg\", where the quote meant to demarcate the end of the
                        // argument gets escaped.
                        if (i == arg.Length - 1)
                        {
                            for (int j = 0; j < slashSequenceLength; j++)
                            {
                                stringBuilder.
                                    Append('\\').
                                    Append('\\');
                            }
                        }
                    }
                    else if (currentChar == '"')
                    {
                        // Every \ or sequence of \ that preceed a " must be escaped.
                        for (int j = 0; j < slashSequenceLength; j++)
                        {
                            stringBuilder.
                                Append('\\').
                                Append('\\');
                        }
                        slashSequenceLength = 0;
    
                        stringBuilder.
                            Append('\\').
                            Append('"');
                    }
                    else
                    {
                        for (int j = 0; j < slashSequenceLength; j++)
                        {
                            stringBuilder.Append('\\');
                        }
                        slashSequenceLength = 0;
    
                        stringBuilder.Append(currentChar);
                    }
                }
    
                return stringBuilder.ToString();
            }
        }
    }
    ```

2. When you add `Jering.Javascript.NodeJS` to your services, overwrite the default `INodeJSProcessFactory` service:
    ```cs
    ServiceCollection services = new ServiceCollection();

    services.AddNodeJS();
    // Override the default service
    services.AddSingleton<INodeJSProcessFactory, CustomNodeJSProcessFactory>();
    ```
   If you're using the static API you'd do it like this:
   ```cs
    ServiceCollection services = new ServiceCollection();

    services.AddNodeJS();
    // Override the default service
    services.AddSingleton<INodeJSProcessFactory, CustomNodeJSProcessFactory>();

   StaticNodeJSService.SetServices(services);

   // Subsequent calls to the static API will use CustomNodeJSProcessFactory
   ```

Let me know if this fixes the issue for you. If it does there's no need for a repro!

**Comment by hansooms999** [comment]
Just got back to this, have not tried it out yet but just set up a sample repo. I will look to see if this fixes it but for now here's the sample repo. https://github.com/hansooms999/JeringConsoleSample/

I'll let you know if the issue is fixed. Thank you!

**Comment by JeremyTCD** [comment]
Just tested your repro with the fix and it works properly now! Will publish a new version `Jering.Javascript.NodeJS` today.

**Comment by JeremyTCD** [comment]
@hansooms999 Fixed version [5.4.4](https://www.nuget.org/packages/Jering.Javascript.NodeJS/5.4.4) 🎉. Thanks for reporting the issue. Let me know if you have any trouble with the new version.

**Comment by hansooms999** [comment]
@JeremyTCD This fixed the issue perfectly. Thanks for your help! 🎉
